nova_ecs/utils.ts

```diff
@@ -47,30 +50,192 @@ export function topologicalSortList(list: Sortable[]): Sortable[] {
+        for (let n of c) {
+            if (domain.has(n)){
+                if (ancestors.has(n)){
+                    const cyclicPart = trimToCyclic(graph);
+                    throw new GraphCycleError("Graph contains cycle\n"+graphListString(cyclicPart));
+                }
+                ancestors.add(n);
+                const next = graph.get(n);
+                if (next) {
+                    dfs(setDifference(next,placed),ancestors);
+                }
+                ancestors.delete(n);
+                result.push(n);
+                placed.add(n);
+            }
+        }
+    }
+    dfs(leafs,new Set<T>());
+    return result;
+}
+
+function graphListString<T>(graph:Map<T, Set<T>>):string{
+    /* produces descriptive text of the style:
+        ┌─c1
+        │ b1
+        │ a1
+       ┌└>root
+       │
+       └──a2
+    */
+    const listing:T[] = [];
+    const indices = new Map<T,number>();
+    const seen = new Set<T>();
+    for (let k of graph.keys()){
+        if (seen.has(k)){
+            continue;
+        }
+        do {
+            seen.add(k);
+            indices.set(k,listing.length);
+            listing.push(k);
+            k = setDifference(graph.get(k)??new Set<T>(),seen).keys().next().value;
+        } while (k)
+    }
+
+    //"│├└┌┼╂┃┠┏┗┞┟↳↱→←"
+    //"│├└┌┼"
+    const cols:string[][] = [];
+    for (let l of listing){
+        const n = graph.get(l);
+        const li = indices.get(l);
+        if (n && (li !== undefined) && n.size > Number(Boolean(listing[li+1] && n.has(listing[li+1])))){
+            const col:string[] = [];
+            let minI = li;
+            let maxI = li;
+            for (let other of n){
+                const i = indices.get(other);
+                if (i !== undefined){
+                    minI = Math.min(i,minI);
+                    maxI = Math.max(i,maxI);
+                }
             }
-            // Check if all nodes this node must come after are already in the list.
-            if (subset(incomingEdges, usedNodes)) {
-                sorted.push(node);
-                usedNodes.add(node);
+            for (let i = 0; i < listing.length; i++){
+                const a = "  ┌┌│├└└  "[(Number(i >= minI) + Number(i > minI) + Number(i >= maxI) + Number(i > maxI))*2+Number(n.has(listing[i]) && i != li+1)]
+                const b = " ><x"[Number(n.has(listing[i]) && i != li+1)+2*Number(i==li)]
+                col.push(a+b);
             }
+            cols.push(col);
         }
-        if (sorted.length === lastLength) {
-            throw new Error('Graph contains a cycle');
+    }
+    let res = "";
+    for (let i = 0; i < listing.length; i++){
+        let line = "";
+        const l = listing[i];
+        const n = graph.get(l);
+        if (!(n && n.has(listing[i+1]))){
+            line = "\x1b[4m";
+        }
+        for (let j = 0; j < cols.length; j++){
+            line += cols[j][i];
+        }
+        res += "\n"+line + l + "\x1b[m";
```

Comment:
Please comment what this escape code does.

Reply:
You could even declare constants like `UNDERLINE` and `CLEAR_FORMATTING` and use them instead. That might be clearer.